docs/zh/config-desc.md
# 配置文件解析

默认在～/.sophon-gateway/config.toml。

```toml
# 被代理线上组件的服务地址
# 可选
Node = "/dns/node/tcp/3453"
Messager = "/dns/messager/tcp/39812"
Droplet = "/dns/market/tcp/41235"
Miner = "/dns/miner/tcp/12308"


[API]
  ListenAddress = "/ip4/127.0.0.1/tcp/45132" # 本地组件wallet和damocles-manager通过长连接和gateway保持通信

[Auth]
  Token = ""
  URL = "http://127.0.0.1:8989"

[Metrics]
  Enabled = false

  [Metrics.Exporter]
    Type = "prometheus" # 两种类型，Graphite或者Prometheus

    [Metrics.Exporter.Graphite]
      Host = "127.0.0.1"
      Namespace = "gateway"
      Port = 4569
      ReportingPeriod = "10s"

    [Metrics.Exporter.Prometheus]
      EndPoint = "/ip4/0.0.0.0/tcp/4569"
      Namespace = "gateway"
      Path = "/debug/metrics"
      RegistryType = "define"
      ReportingPeriod = "10s"

[RateLimit]
  #redis地址，用于记录用户访问的次数。如果要开启对某个user的访问限速，还需要`auth` 服务同时设置`sophon-auth user rate-limit`命令。
  Redis = "27.0.0.1:6379" 

[Trace]
  JaegerEndpoint = "localhost:6831"
  JaegerTracingEnabled = false
  ProbabilitySampler = 1.0
  ServerName = "sophon-gateway"

```


docs/zh/metrics-config.md
# metrics 配置及使用说明

## 配置

`Metrics` 基本的配置样例如下：
```toml
[Metrics]
  # 是否开启metrics指标统计，默认为false
  Enabled = false
  
  [Metrics.Exporter]
    # 指标导出器类型，目前可选：prometheus或graphite，默认为prometheus
    Type = "prometheus"
    
    [Metrics.Exporter.Prometheus]
      # multiaddr
      EndPoint = "/ip4/0.0.0.0/tcp/4569"
      # 命名规范: "a_b_c", 不能带"-"
      Namespace = "messager01" 
      # 指标注册表类型，可选：default（默认，会附带程序运行的环境指标）或 define（自定义）
      RegistryType = "define"
      # prometheus 服务路径
      Path = "/debug/metrics"
      # 上报周期
      ReportingPeriod = "10s"
      
    [Metrics.Exporter.Graphite]
      # 命名规范: "a_b_c", 不能带"-"
      Namespace = "messager01" 
      # graphite exporter 收集器服务地址
      Host = "127.0.0.1"
      # graphite exporter 收集器服务监听端口
      Port = 4569
      # 上报周期
      ReportingPeriod = "10s"
```
## 导出器

目前可以选择两类导出器（`exporter`）：`Prometheus exporter` 或 `Graphite exporter`，默认是前者。

如果配置 `Prometheus exporter`，则在 `venus-messager` 服务启动时会附带启动 `Prometheus exporter` 的监听服务，可以通过以下方式快速查看指标：


```bash
$ curl http://localhost:4569/debug/metrics
  gateway_wallet_sign_bucket{wallet_account="forcenet-nv16",le="0.01"} 0
  gateway_wallet_sign_bucket{wallet_account="forcenet-nv16",le="0.05"} 0
  gateway_wallet_sign_bucket{wallet_account="forcenet-nv16",le="0.1"} 0
  gateway_wallet_sign_bucket{wallet_account="forcenet-nv16",le="0.3"} 0
  gateway_wallet_sign_bucket{wallet_account="forcenet-nv16",le="0.6"} 0
  gateway_wallet_sign_bucket{wallet_account="forcenet-nv16",le="0.8"} 0
  gateway_wallet_sign_bucket{wallet_account="forcenet-nv16",le="1"} 0
  gateway_wallet_sign_bucket{wallet_account="forcenet-nv16",le="2"} 322
  gateway_wallet_sign_bucket{wallet_account="forcenet-nv16",le="3"} 544
  gateway_wallet_sign_bucket{wallet_account="forcenet-nv16",le="4"} 787
  gateway_wallet_sign_bucket{wallet_account="forcenet-nv16",le="5"} 790
  gateway_wallet_sign_bucket{wallet_account="forcenet-nv16",le="6"} 791
  ... ...
```
> 如果遇到错误 `curl: (56) Recv failure: Connection reset by peer`, 请使用本机 `ip` 地址, 如下所示:
```bash
$  curl http://<ip>:4569/debug/metrics
```

如果配置 `Graphite exporter`，需要先启动 `Graphite exporter` 的收集器服务， `venus-messager` 服务启动时将指标上报给收集器。服务启动参考 [Graphite exporter](https://github.com/prometheus/graphite_exporter) 中的说明。

`Graphite exporter` 和 `Prometheus exporter` 自身都不带图形界面的，如果需要可视化监控及更高阶的图表分析，请到 `venus-docs` 项目中查找关于 `Prometheus+Grafana` 的说明文档。


CHANGELOG.md
# sophon-gateway changelog

## v1.12.0

* feat: move RandomBytes to venus share by @LinZexiao in [[#141](https://github.com/ipfs-force-community/sophon-gateway/pull/141)]
* doc: fix link by @0x5459 in [[#139](https://github.com/ipfs-force-community/sophon-gateway/pull/139)]

## v1.12.0-rc1

* chore: merge release/v1.11 into master [[#131](https://github.com/ipfs-force-community/sophon-gateway/pull/131)]
* chore: upgrade venus and venus-auth by [[#132](https://github.com/ipfs-force-community/sophon-gateway/pull/132)]
* Feat/tanlang/add unseal state[[#133](https://github.com/ipfs-force-community/sophon-gateway/pull/133)]
* feat: replace venus-gateway with sophon-gateway by [[#135](https://github.com/ipfs-force-community/sophon-gateway/pull/135)]
* build(deps): bump github.com/gin-gonic/gin from 1.9.0 to 1.9.1 [[#134](https://github.com/ipfs-force-community/sophon-gateway/pull/134)]

## v1.11.1

* bump up ffi
## v1.11.0

* bump up version to v1.11.0

## v1.11.0-rc1

### New Features
* feat: add status api to detect api ready  / 添加状态检测接口[[#116](https://github.com/ipfs-force-community/sophon-gateway/pull/116)]
* feat:update the auth client with token  /客户端token验证 [[#117](https://github.com/ipfs-force-community/sophon-gateway/pull/117)]
* feat: refactor market event / 重构market event [[#112](https://github.com/ipfs-force-community/sophon-gateway/pull/112)]
* feat: rm isUnseal method in market envent / 移除 IsUnseal 方法  [[#125](https://github.com/ipfs-force-community/sophon-gateway/pull/125)]
* feat: add docker push /增加推送到镜像仓库的功能 [[#126](https://github.com/ipfs-force-community/sophon-gateway/pull/126)]

# v1.10.1

* 支持注册 delegated address [[#122](https://github.com/ipfs-force-community/sophon-gateway/pull/122)]
* 升级 venus 和 sophon-auth 版本到 v1.10.1

## v1.10.0

* 升级 venus 和 sophon-auth 版本到 v1.10.0
* 升级 filecoin-ffi 版本到 v1.20.0
* 升级 go-jsonrpc 版本到 v0.1.7


config/config.go
package config

import (
	"io/ioutil"

	"github.com/ipfs-force-community/metrics"
	"github.com/pelletier/go-toml"
)

const (
	// Configuration file name
	ConfigFile = "config.toml"
)

type Config struct {
	API  *APIConfig
	Auth *AuthConfig

	Node     *string
	Messager *string
	Miner    *string
	Droplet  *string

	Metrics   *metrics.MetricsConfig
	Trace     *metrics.TraceConfig
	RateLimit *RateLimitCofnig
}

type APIConfig struct {
	ListenAddress string
}

type AuthConfig struct {
	URL   string
	Token string
}

type RateLimitCofnig struct {
	Redis string
}

func DefaultConfig() *Config {
	cfg := &Config{
		API:       &APIConfig{ListenAddress: "/ip4/127.0.0.1/tcp/45132"},
		Auth:      &AuthConfig{URL: "http://127.0.0.1:8989"},
		Metrics:   metrics.DefaultMetricsConfig(),
		Trace:     metrics.DefaultTraceConfig(),
		RateLimit: &RateLimitCofnig{Redis: ""},
	}
	namespace := "gateway"
	cfg.Metrics.Exporter.Prometheus.Namespace = namespace
	cfg.Metrics.Exporter.Graphite.Namespace = namespace
	cfg.Metrics.Exporter.Prometheus.EndPoint = "/ip4/0.0.0.0/tcp/4569"
	cfg.Metrics.Exporter.Graphite.Port = 4569
	cfg.Trace.ServerName = "sophon-gateway"
	cfg.Trace.JaegerEndpoint = ""

	return cfg
}

func ReadConfig(filePath string) (*Config, error) {
	data, err := ioutil.ReadFile(filePath)
	if err != nil {
		return nil, err
	}

	cfg := &Config{}
	err = toml.Unmarshal(data, cfg)

	return cfg, err
}

func WriteConfig(filePath string, cfg *Config) error {
	data, err := toml.Marshal(cfg)
	if err != nil {
		return err
	}

	return ioutil.WriteFile(filePath, data, 0o644)
}
